### PR TITLE
feat(makeChainInfo): optional apiAddrs

### PR DESCRIPTION
**Pull request by turadg**
## Description

Follow up to https://github.com/Agoric/agoric-sdk/pull/6504

- adds static type
- adds tests
- makes it fall back to older behavior when `apiAddrs` absent in network-config

### Security Considerations

.
### Documentation Considerations

.
### Testing Considerations

unit

### CODE REVIEW

File: packages/wallet/ui/src/util/SuggestChain.js [hunk 1]
```diff
@@ -5,36 +5,33 @@ export const AGORIC_COIN_TYPE = 564;
...
-  caption,
-) => {
-  const coinType = Number(
-    new URL(networkConfig).searchParams.get('coinType') || AGORIC_COIN_TYPE,
-  );
-  const hostname = new URL(networkConfig).hostname;
-  const network = hostname.split('.')[0];
+export const makeChainInfo = (io, networkConfig, caption) => {
+  const { chainName, rpcAddrs, apiAddrs } = networkConfig;
+  const index = Math.floor(io.random() * rpcAddrs.length);
```
Comment (dckc) [inline packages/wallet/ui/src/util/SuggestChain.js]: passing in the random number generator for 1 call seems a little odd -- why not just pass in the results of the call?

File: packages/wallet/ui/src/util/tests/SuggestChain.test.js [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+ */
+const ollinetExample = {
+  chainName: 'agoricollinet-44',
+  gci: 'https://ollinet.rpc.agoric.net:443/genesis',
+  peers: ['fb86a0993c694c981a28fa1ebd1fd692f345348b@34.67.167.8:26656'],
+  rpcAddrs: ['https://ollinet.rpc.agoric.net:443'],
+  seeds: ['0f04c4610b7511a64b8644944b907416db568590@34.172.146.145:26656'],
+};
+
+const io = { random: Math.random };
```
Comment (dckc) [inline packages/wallet/ui/src/util/tests/SuggestChain.test.js]: We have only 1 entry in our `rpcAddrs` array, so it doesn't matter yet, but this should be a mock random number generator with predictable outputs for testing.

File: packages/wallet/ui/src/util/tests/SuggestChain.test.js [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+
+test('mainnetExample', () => {
```
Comment (dckc) [inline packages/wallet/ui/src/util/tests/SuggestChain.test.js]: I suppose this covers it, but what I had in mind was a test assertion of the form "rpcAddr parameter is not clobbered".

File: packages/wallet/ui/src/util/SuggestChain.js [hunk 4]
```diff
@@ -5,36 +5,33 @@ export const AGORIC_COIN_TYPE = 564;
...
   return {
     rpc,
-    rest: apiAddr,
-    chainId,
-    chainName: caption || `Agoric ${network}`,
+    rest,
+    chainId: chainName,
+    chainName: caption,
     stakeCurrency,
-    walletUrlForStaking: `https://${network}.staking.agoric.app`,
```
Comment (dckc) [inline packages/wallet/ui/src/util/SuggestChain.js:35]: No more staking link? That looks like a regression / breaking change. I have used it from time to time. I suppose I won't miss it that badly.

File: packages/wallet/ui/src/util/SuggestChain.js [hunk 5]
```diff
@@ -44,41 +41,38 @@ const makeChainInfo = (
...
+  const res = await fetch(url);
   if (!res.ok) {
     throw Error(`Cannot fetch network: ${res.status}`);
   }
 
-  const { chainName: chainId, apiAddrs, rpcAddrs } = await res.json();
-  assert.equal(apiAddrs.length, rpcAddrs.length);
-  const index = Math.floor(random() * rpcAddrs.length);
+  /** @type {import('@agoric/internal').NetworkConfig} */
+  const networkConfig = await res.json();
```
Comment (dckc) [inline packages/wallet/ui/src/util/SuggestChain.js]: I make this sort of type annotation all the time when I'm prototyping, but it's unsound. I suppose it's just documenting an assumption we were already making, so it's not a regression. But we should add a shape guard or the like. Or use type `unknown` so tsc forces us to validate each assumption we make about this input.

Please make some XXX annotation to say "this is not best practice; don't copy this idiom blindly".

File: packages/wallet/ui/src/util/SuggestChain.js (hunk 4 above)
Comment (dckc) [reply packages/wallet/ui/src/util/SuggestChain.js:35]: Hm... we did ship, so maybe it's worth more discipline about this sort of thing. For the benefit of release notes, how about a separate BREAKING CHANGE commit to take out `walletUrlForStaking` before this `feat`?

**Review by dckc — COMMENTED** [review]
I'd like to discuss a couple things before approving.

File: packages/wallet/ui/src/util/SuggestChain.js (hunk 1 above)
Comment (turadg) [reply packages/wallet/ui/src/util/SuggestChain.js]: agreed. the dan voice in my head said, "pass the ocap" but that Dan is an crinkled facsimile of the real one

File: packages/wallet/ui/src/util/SuggestChain.js (hunk 5 above)
Comment (turadg) [reply packages/wallet/ui/src/util/SuggestChain.js]: I think this is the order of bad to good:
1. implicit `any` (what we have all over the place)
2. constrain `any` (what is here)
3. unknown (so you have to validate on use, but a pain)
4. type guard (which is why I put the TODO for that)

I'm having trouble thinking up what the unsound type exploit is here. If someone controls network config, there are much worse things they can do

File: packages/wallet/ui/src/util/SuggestChain.js (hunk 4 above)
Comment (turadg) [reply packages/wallet/ui/src/util/SuggestChain.js:35]: I found it nowhere in the Agoric org (https://github.com/Agoric/agoric-sdk/search?q=walletUrlForStaking) and I didn't think it was really part of the API of this util.

It's technically breaking the API, but I don't think it breaks anything. If we're going to get disciplined about breaking changes, then I think we also need to get more disciplined about what we export.

In any case, I'll try separating this out as a BREAKING CHANGE to see how it feels ;-)

Comment (dckc) [reply packages/wallet/ui/src/util/SuggestChain.js:35]: It's not used in our code; it's used in by Keplr:

image


Again: this is a network protocol / API.

File: packages/wallet/ui/src/util/SuggestChain.js (hunk 5 above)
Comment (dckc) [reply packages/wallet/ui/src/util/SuggestChain.js]: Agreed on the order. I guess I was reading too fast and missed the TODO.

> I'm having trouble thinking up what the unsound type exploit is here.

I don't know that this is exploitable. But the unsoundness adds a burden when reviewing; I have to keep track of all the places where `networkConfig` is used, knowing that it's really an external input and might not be of the declared type.

Thanks for the type guard.

**Review by dckc — APPROVED** [review]

File: packages/wallet/ui/src/util/SuggestChain.js (hunk 4 above)
Comment (turadg) [reply packages/wallet/ui/src/util/SuggestChain.js:35]: Ah. I'll restore it.

(Fwiw, I don't see the antecedent for "again". It definitely wasn't clear from your earlier comments that this was an actual regression. )

**Comment by turadg** [comment]
@dckc I'm putting this back into draft until I can document [gci, peers, seeds](https://github.com/Agoric/agoric-sdk/pull/6505/files#diff-be19ecccf4827e08edb9e57079dd7b3dabd0c37ea9fce2874b62b12e2dc97a80R10-R12)

**Review by dckc — APPROVED** [review]
LGTM

File: packages/casting/src/netconfig.js [hunk 6]
```diff
@@ -9,19 +9,19 @@ import { fit, M } from '@agoric/store';
  * “light clients” to prove the ancestry of data fetched from the rpc node.
  *
  * @property {string[]} rpcAddrs endpoints serving https://docs.tendermint.com/v0.34/rpc/
- * @property {string} [chainName] e.g. agoric-3
+ * @property {string} chainName e.g. agoric-3
```
Comment (dckc) [inline packages/casting/src/netconfig.js:12]: oops; I wasn't reviewing very carefully, was I?

Comment (turadg) [reply packages/casting/src/netconfig.js:12]: type system thing, rare for human eyes to catch ime

**Pull request merged**